mitmproxy/addons/share.py

```diff
@@ -0,0 +1,79 @@
+import typing
+import random
+import datetime
+import time
+import _io
+import http.client
+
+from mitmproxy import command
+from mitmproxy import io
+from mitmproxy import ctx
+from mitmproxy import flow
+
+
+class Share:
+    def __init__(self):
+        self.res = None
+        self.u_id = None
+
+    def encode_multipart_formdata(self, filename, content):
+        params = {"key": filename, "acl": "bucket-owner-full-control", "Content-Type": "application/octet-stream"}
+        LIMIT = b'---------------------------198495659117975628761412556003'
+        CRLF = b'\r\n'
+        l = []
+        for (key, value) in params.items():
+            l.append(b'--' + LIMIT)
+            l.append(b'Content-Disposition: form-data; name="%b"' % key.encode("utf-8"))
+            l.append(b'')
+            l.append(value.encode("utf-8"))
+        l.append(b'--' + LIMIT)
+        l.append(b'Content-Disposition: form-data; name="file"; filename="%b"' % filename.encode("utf-8"))
+        l.append(b'Content-Type: application/octet-stream')
+        l.append(b'')
+        l.append(content)
+        l.append(b'--' + LIMIT + b'--')
+        l.append(b'')
+        body = CRLF.join(l)
+        content_type = b'multipart/form-data; boundary=%b' % LIMIT
+        return content_type, body
```

Comment:
We already depend on the `requests` package, why not use its support for multipart uploads? http://docs.python-requests.org/en/master/user/quickstart/#post-a-multipart-encoded-file

Reply:
Apart from `requests`, there was `urllib` also implemented in some of the files but @mhils recommended to use http.client.

Reply:
Yes, we use `urllib` only for low-level parsing of HTTP-related things.
But in this case (upload files to a server), I think using `requests` is the better and easier choice.

Reply:
So we are having our own `multipart`?  :)

Reply:
Yes - lets keep it.